### PR TITLE
[KYUUBI #1105] Zorder codegen support

### PR DESCRIPTION
**Pull request by cfmcgrady**
### _Why are the changes needed?_

1. zorder codegen support
2. move all of the zorder tests to a separate file.
3. add more zorder unit tests.

generate.java
```java
/* 001 */ public java.lang.Object generate(Object[] references) {
/* 002 */   return new SpecificUnsafeProjection(references);
/* 003 */ }
/* 004 */
/* 005 */ class SpecificUnsafeProjection extends org.apache.spark.sql.catalyst.expressions.UnsafeProjection {
/* 006 */
/* 007 */   private Object[] references;
/* 008 */   private org.apache.spark.sql.catalyst.expressions.codegen.UnsafeRowWriter[] mutableStateArray_0 = new org.apache.spark.sql.catalyst.expressions.codegen.UnsafeRowWriter[1];
/* 009 */
/* 010 */   public SpecificUnsafeProjection(Object[] references) {
/* 011 */     this.references = references;
/* 012 */     mutableStateArray_0[0] = new org.apache.spark.sql.catalyst.expressions.codegen.UnsafeRowWriter(1, 32);
/* 013 */
/* 014 */   }
/* 015 */
/* 016 */   public void initialize(int partitionIndex) {
/* 017 */
/* 018 */   }
/* 019 */
/* 020 */   // Scala.Function1 need this
/* 021 */   public java.lang.Object apply(java.lang.Object row) {
/* 022 */     return apply((InternalRow) row);
/* 023 */   }
/* 024 */
/* 025 */   public UnsafeRow apply(InternalRow i) {
/* 026 */     mutableStateArray_0[0].reset();
/* 027 */
/* 028 */
/* 029 */
/* 030 */
/* 031 */     byte[] value_0 = null;
/* 032 */     byte[][] binaryArray_0 = new byte[2][];
/* 033 */
/* 034 */     boolean isNull_1 = i.isNullAt(0);
/* 035 */     long value_1 = isNull_1 ?
/* 036 */     -1L : (i.getLong(0));
/* 037 */     if (isNull_1) {
/* 038 */       binaryArray_0[0] = (byte[]) ((byte[][]) references[0] /* defaultValues */)[0];
/* 039 */     } else {
/* 040 */       binaryArray_0[0] = org.apache.kyuubi.sql.zorder.ZorderBytesUtils.toByte(value_1);
/* 041 */     }
/* 042 */
/* 043 */
/* 044 */     boolean isNull_2 = i.isNullAt(1);
/* 045 */     long value_2 = isNull_2 ?
/* 046 */     -1L : (i.getLong(1));
/* 047 */     if (isNull_2) {
/* 048 */       binaryArray_0[1] = (byte[]) ((byte[][]) references[0] /* defaultValues */)[1];
/* 049 */     } else {
/* 050 */       binaryArray_0[1] = org.apache.kyuubi.sql.zorder.ZorderBytesUtils.toByte(value_2);
/* 051 */     }
/* 052 */
/* 053 */     value_0 = org.apache.kyuubi.sql.zorder.ZorderBytesUtils.interleaveMultiByteArray(binaryArray_0);
/* 054 */     mutableStateArray_0[0].write(0, value_0);
/* 055 */     return (mutableStateArray_0[0].getRow());
/* 056 */   }
/* 057 */
/* 058 */
/* 059 */ }
```

### _How was this patch tested?_
- [x] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [ ] [Run test](https://kyuubi.readthedocs.io/en/latest/develop_tools/testing.html#running-tests) locally before make a pull request

### CODE REVIEW

File: dev/kyuubi-extension-spark-3-1/src/main/scala/org/apache/kyuubi/sql/zorder/Zorder.scala [hunk 1]
```diff
@@ -42,21 +46,61 @@ case class Zorder(children: Seq[Expression]) extends Expression with CodegenFall
...
+    ZorderBytesUtils.interleaveMultiByteArray(binaryArr.toArray)
+  }
 
-    val binaryArr = evaluated.map(ZorderBytesUtils.toByte).toArray
-    ZorderBytesUtils.interleaveMultiByteArray(binaryArr)
+  override protected def doGenCode(ctx: CodegenContext, ev: ExprCode): ExprCode = {
+    val evals = children.map(_.genCode(ctx))
+    val defaultValues = ctx.addReferenceObj("defaultValues", defaultNullValues)
+    val binaryArray = ctx.freshName("binaryArray")
+    val util = ZorderBytesUtils.getClass.getName + ".MODULE$"
```
Comment (yaooqinn) [inline dev/kyuubi-extension-spark-3-1/src/main/scala/org/apache/kyuubi/sql/zorder/Zorder.scala:77]: ZorderBytesUtils.getClass.getName.stripSuffix("$")

File: pom.xml [hunk 2]
```diff
@@ -405,6 +405,13 @@
                 </exclusions>
             </dependency>
 
+            <dependency>
+                <groupId>org.scalatestplus</groupId>
+                <artifactId>scalatestplus-scalacheck_${scala.binary.version}</artifactId>
```
Comment (cfmcgrady) [inline pom.xml:410]: `ExpressionEvalHelper` need this one.

File: dev/kyuubi-extension-spark-3-1/src/main/scala/org/apache/kyuubi/sql/zorder/Zorder.scala (hunk 1 above)
Comment (cfmcgrady) [reply dev/kyuubi-extension-spark-3-1/src/main/scala/org/apache/kyuubi/sql/zorder/Zorder.scala]: Done

File: dev/kyuubi-extension-spark-3-1/src/test/scala/org/apache/spark/sql/ZorderSuite.scala [hunk 3]
```diff
@@ -0,0 +1,456 @@
...
+      0x44, 0x44, 0x22, 0x22, 0x08, 0x88, 0x82, 0x22, 0x20, 0x88,
+      0x88, 0x22, 0x22, 0x08, 0x88, 0x82, 0x22, 0x20, 0xa8, 0x9a,
+      0x2a, 0x2a, 0x8a, 0x8a, 0xa2, 0xa2, 0xa8, 0xa8, 0xaa, 0x2a,
+      0x2a, 0x8a, 0x8a, 0xa2, 0xa2, 0xa8, 0xa8, 0xaa, 0xca, 0x8a,
+      0xa8, 0xa8, 0xaa, 0x8a, 0x8a, 0xa8, 0xa8, 0xaa, 0x8a, 0x8a,
+      0xa8, 0xa8, 0xaa, 0x8a, 0x8a, 0xa8, 0xa8, 0xaa, 0xb2, 0xa2,
+      0xaa, 0x8a, 0x9a, 0xaa, 0x2a, 0x6a, 0xa8, 0xa8, 0xaa, 0xa2,
+      0xa2, 0xaa, 0x8a, 0x8a, 0xaa, 0x2f, 0x6b, 0xfc)
+      .map(_.toByte)
+    checkEvaluation(zorder, expected, InternalRow.fromSeq(children))
```
Comment (cfmcgrady) [inline dev/kyuubi-extension-spark-3-1/src/test/scala/org/apache/spark/sql/ZorderSuite.scala:390]: use `org.apache.spark.sql.catalyst.expressions.ExpressionEvalHelper.checkEvaluation()` to do assumptions as it has more test cases.

**Comment by cfmcgrady** [comment]
cc @ulysses-you @yaooqinn @qphien

File: dev/kyuubi-extension-spark-3-1/src/main/scala/org/apache/kyuubi/sql/zorder/Zorder.scala [hunk 4]
```diff
@@ -42,21 +46,61 @@ case class Zorder(children: Seq[Expression]) extends Expression with CodegenFall
   }
 
   @transient
-  private lazy val defaultNullValues: Seq[Any] = {
-    children.map(child => ZorderBytesUtils.defaultValue(child.dataType))
+  private lazy val defaultNullValues: JMap[Int, Array[Byte]] = {
```
Comment (yaooqinn) [inline dev/kyuubi-extension-spark-3-1/src/main/scala/org/apache/kyuubi/sql/zorder/Zorder.scala:49]: Using idx as a key seems not necessary? we can use an arrary/seq here ?

**Comment by yaooqinn** [comment]
can we print the `generate.java` here in PR desc or somewhere else in the comment zone?

**Comment by yaooqinn** [comment]
cc @ulysses-you @pan3793

File: dev/kyuubi-extension-spark-3-1/src/test/scala/org/apache/spark/sql/ZorderSuite.scala (hunk 3 above)
Comment (ulysses-you) [reply dev/kyuubi-extension-spark-3-1/src/test/scala/org/apache/spark/sql/ZorderSuite.scala:390]: Can we add a codgen result test here to check the byte array ?. Something like:

```
val query =
  """
    |SELECT zorder(
    |  false,
    |  cast(null as boolean),
    |  cast(1 as byte),
    |  cast(null as byte),
    |  cast(1 as short),
    |  cast(null as short),
    |  1,
    |  cast(null as int),
    |  1L,
    |  cast(null as long),
    |  1f,
    |  cast(null as float),
    |  1d,
    |  cast(null as double),
    |  "1",
    |  cast(null as string),
    |  cast(1L as long),
    |  cast(null as timestamp),
    |  cast(1 as int),
    |  cast(null as date),
    |  cast(0.1 as decimal(1,1)),
    |  cast(null as decimal(1,1))
    |)
    |""".stripMargin
Seq(true, false).foreach { mode =>
  withSQLConf(SQLConf.WHOLESTAGE_CODEGEN_ENABLED.key -> mode.toString) {
    checkAnswer(sql(query), Row(expected))
  }
}
```

File: dev/kyuubi-extension-spark-3-1/src/test/scala/org/apache/spark/sql/KyuubiExtensionSuite.scala [hunk 5]
```diff
@@ -62,6 +59,8 @@ class KyuubiExtensionSuite extends QueryTest with SQLTestUtils with AdaptiveSpar
     if (_spark != null) {
       _spark.stop()
     }
+    Utils.deleteRecursively(new java.io.File("spark-warehouse"))
+    Utils.deleteRecursively(new java.io.File("metastore_db"))
```
Comment (ulysses-you) [inline dev/kyuubi-extension-spark-3-1/src/test/scala/org/apache/spark/sql/KyuubiExtensionSuite.scala:63]: I see your requirements. We can share the one spark session between all test in extension, so we don't need do this one by one.

File: pom.xml [hunk 6]
```diff
@@ -405,6 +405,13 @@
...
+                <version>3.1.0.0-RC2</version>
```
Comment (ulysses-you) [inline pom.xml:411]: cc @pan3793

**Review by ulysses-you — COMMENTED** [review]
thank you @cfmcgrady for the work !

File: dev/kyuubi-extension-spark-3-1/src/test/scala/org/apache/spark/sql/ZorderSuite.scala (hunk 3 above)
Comment (cfmcgrady) [reply dev/kyuubi-extension-spark-3-1/src/test/scala/org/apache/spark/sql/ZorderSuite.scala:390]: I think `checkEvaluation()` already does this?

File: dev/kyuubi-extension-spark-3-1/src/main/scala/org/apache/kyuubi/sql/zorder/Zorder.scala (hunk 4 above)
Comment (cfmcgrady) [reply dev/kyuubi-extension-spark-3-1/src/main/scala/org/apache/kyuubi/sql/zorder/Zorder.scala:49]: Yes, remove the sort and update to seq.

**Comment by cfmcgrady** [comment]
> can we print the generate.java here in PR desc or somewhere else in the comment zone?

Updated the PR desc.

File: dev/kyuubi-extension-spark-3-1/src/main/scala/org/apache/kyuubi/sql/zorder/Zorder.scala [hunk 7]
```diff
@@ -42,21 +46,61 @@ case class Zorder(children: Seq[Expression]) extends Expression with CodegenFall
...
+      val inputs = evals.zipWithIndex.map {
+        case (eval, index) =>
+          code"""
+              |${eval.code}
+              |if (${eval.isNull}) {
+              |  $binaryArray[$index] = (byte[]) $defaultValues.get($index);
+              |} else {
+              |  $binaryArray[$index] = $util.toByte(
+              |    ${eval.value}
+              |  );
```
[the remark this answers is not in the document]
Comment (cfmcgrady) [reply dev/kyuubi-extension-spark-3-1/src/main/scala/org/apache/kyuubi/sql/zorder/Zorder.scala]: done.

File: dev/kyuubi-extension-spark-3-1/src/main/scala/org/apache/kyuubi/sql/zorder/Zorder.scala [hunk 8]
```diff
@@ -42,21 +46,61 @@ case class Zorder(children: Seq[Expression]) extends Expression with CodegenFall
...
-    ZorderBytesUtils.interleaveMultiByteArray(binaryArr)
+  override protected def doGenCode(ctx: CodegenContext, ev: ExprCode): ExprCode = {
+    val evals = children.map(_.genCode(ctx))
+    val defaultValues = ctx.addReferenceObj("defaultValues", defaultNullValues)
+    val binaryArray = ctx.freshName("binaryArray")
+    val util = ZorderBytesUtils.getClass.getName.stripSuffix("$")
+    try {
+      val inputs = evals.zipWithIndex.map {
+        case (eval, index) =>
+          code"""
```
[the remark this answers is not in the document]
Comment (cfmcgrady) [reply dev/kyuubi-extension-spark-3-1/src/main/scala/org/apache/kyuubi/sql/zorder/Zorder.scala]: done.

File: dev/kyuubi-extension-spark-3-1/src/main/scala/org/apache/kyuubi/sql/zorder/Zorder.scala [hunk 9]
```diff
@@ -42,21 +46,61 @@ case class Zorder(children: Seq[Expression]) extends Expression with CodegenFall
...
+              |}
+              |""".stripMargin
+      }
+      ev.copy(code =
+        code"""
+            |byte[] ${ev.value} = null;
+            |byte[][] $binaryArray = new byte[${evals.length}][];
+            |${inputs.mkString("\n")}
+            |${ev.value} = $util.interleaveMultiByteArray($binaryArray);
+            |boolean ${ev.isNull} = ${ev.value} == null;
```
[the remark this answers is not in the document]
Comment (cfmcgrady) [reply dev/kyuubi-extension-spark-3-1/src/main/scala/org/apache/kyuubi/sql/zorder/Zorder.scala]: good point!

File: dev/kyuubi-extension-spark-3-1/src/main/scala/org/apache/kyuubi/sql/zorder/Zorder.scala [hunk 10]
```diff
@@ -42,21 +46,61 @@ case class Zorder(children: Seq[Expression]) extends Expression with CodegenFall
...
+            |""".stripMargin)
+    } catch {
```
[the remark this answers is not in the document]
Comment (cfmcgrady) [reply dev/kyuubi-extension-spark-3-1/src/main/scala/org/apache/kyuubi/sql/zorder/Zorder.scala]: It's unnecessary, updated.

File: dev/kyuubi-extension-spark-3-1/src/main/scala/org/apache/kyuubi/sql/zorder/Zorder.scala [hunk 11]
```diff
@@ -17,15 +17,19 @@
...
-import org.apache.spark.sql.catalyst.expressions.codegen.CodegenFallback
+import org.apache.spark.sql.catalyst.expressions.{ComplexTypeMergingExpression, Expression}
+import org.apache.spark.sql.catalyst.expressions.codegen.{CodegenContext, ExprCode, FalseLiteral}
+import org.apache.spark.sql.catalyst.expressions.codegen.Block._
 import org.apache.spark.sql.types.{BinaryType, DataType}
 
 import org.apache.kyuubi.sql.KyuubiSQLExtensionException
 
-case class Zorder(children: Seq[Expression]) extends Expression with CodegenFallback {
+case class Zorder(children: Seq[Expression]) extends ComplexTypeMergingExpression with Logging {
```
Comment (ulysses-you) [inline dev/kyuubi-extension-spark-3-1/src/main/scala/org/apache/kyuubi/sql/zorder/Zorder.scala:32]: does `Expression` not good enough ? why we need `ComplexTypeMergingExpression` ?

File: dev/kyuubi-extension-spark-3-1/src/test/scala/org/apache/spark/sql/ZorderSuite.scala (hunk 3 above)
Comment (ulysses-you) [reply dev/kyuubi-extension-spark-3-1/src/test/scala/org/apache/spark/sql/ZorderSuite.scala:390]: ah I see, `checkEvaluation` is good enough

File: dev/kyuubi-extension-spark-3-1/src/main/scala/org/apache/kyuubi/sql/zorder/Zorder.scala (hunk 11 above)
Comment (cfmcgrady) [reply dev/kyuubi-extension-spark-3-1/src/main/scala/org/apache/kyuubi/sql/zorder/Zorder.scala]: `Expression` is good enough, reverted it.

File: dev/kyuubi-extension-spark-3-1/src/test/scala/org/apache/spark/sql/KyuubiExtensionSuite.scala (hunk 5 above)
Comment (cfmcgrady) [reply dev/kyuubi-extension-spark-3-1/src/test/scala/org/apache/spark/sql/KyuubiExtensionSuite.scala:63]: What can we do if we want to share a single session between two `Suites`?

File: dev/kyuubi-extension-spark-3-1/src/main/scala/org/apache/kyuubi/sql/zorder/Zorder.scala (hunk 4 above)
Comment (ulysses-you) [reply dev/kyuubi-extension-spark-3-1/src/main/scala/org/apache/kyuubi/sql/zorder/Zorder.scala:48]: I think the comment of @yaooqinn is saying: `private lazy val defaultNullValues: Array[Array[Byte]]`, we don't need map here.

File: dev/kyuubi-extension-spark-3-1/src/test/scala/org/apache/spark/sql/KyuubiExtensionSuite.scala (hunk 5 above)
Comment (cfmcgrady) [reply dev/kyuubi-extension-spark-3-1/src/test/scala/org/apache/spark/sql/KyuubiExtensionSuite.scala:63]: I think I see.

Comment (ulysses-you) [reply dev/kyuubi-extension-spark-3-1/src/test/scala/org/apache/spark/sql/KyuubiExtensionSuite.scala:63]: Use a singleton class to init spark session ?

File: dev/kyuubi-extension-spark-3-1/src/main/scala/org/apache/kyuubi/sql/zorder/Zorder.scala (hunk 4 above)
Comment (cfmcgrady) [reply dev/kyuubi-extension-spark-3-1/src/main/scala/org/apache/kyuubi/sql/zorder/Zorder.scala]: Sorry, I misread before, using an array here is ok, reverted it.

File: pom.xml (hunk 6 above)
Comment (pan3793) [reply pom.xml:411]: Can we switch to 
```
<dependency>
    <groupId>org.scalatestplus</groupId>
    <artifactId>scalacheck-1-15_2.12</artifactId>
    <version>3.2.9.0</version>
    <scope>test</scope>
</dependency>
```
?

**Review by ulysses-you — APPROVED** [review]
LGTM, the [comment](https://github.com/apache/incubator-kyuubi/pull/1109#discussion_r709838217) we can do it in a new PR.

File: pom.xml (hunk 6 above)
Comment (pan3793) [reply pom.xml:411]: BTW, please always extract `version` as property.

**Comment by cfmcgrady** [comment]
> LGTM, the [comment](https://github.com/apache/incubator-kyuubi/pull/1109#discussion_r709838217) we can do it in a new PR.

Ok, I will raise a new PR after this PR  be merged.

File: pom.xml (hunk 6 above)
Comment (cfmcgrady) [reply pom.xml]: Updated.

**Comment by ulysses-you** [comment]
thank you @cfmcgrady , merged to master

File: dev/kyuubi-extension-spark-3-1/src/main/scala/org/apache/kyuubi/sql/zorder/Zorder.scala [hunk 12]
```diff
@@ -42,21 +43,48 @@ case class Zorder(children: Seq[Expression]) extends Expression with CodegenFall
   }
 
   @transient
-  private lazy val defaultNullValues: Seq[Any] = {
-    children.map(child => ZorderBytesUtils.defaultValue(child.dataType))
-  }
+  private lazy val defaultNullValues: Array[Array[Byte]] =
+    children.map(_.dataType)
+      .map(ZorderBytesUtils.defaultValue)
+      .map(ZorderBytesUtils.toByte)
```
Comment (ulysses-you) [inline dev/kyuubi-extension-spark-3-1/src/main/scala/org/apache/kyuubi/sql/zorder/Zorder.scala:49]: @cfmcgrady we can simplify `ZorderBytesUtils.defaultValue` to return `Array[Byte]` directly. So we don't need call .map twice.

Comment (cfmcgrady) [reply dev/kyuubi-extension-spark-3-1/src/main/scala/org/apache/kyuubi/sql/zorder/Zorder.scala:49]: Good catch, will raise a new pr.

**Pull request closed without merging**